lib/spells/effects/Catapult.cpp

```diff
@@ -97,29 +85,121 @@ void Catapult::apply(ServerCallback * server, const Mechanics * m, const EffectT
 		// Potential overshots (more hits on same targets than remaining HP) are allowed
 		EWallPart target = *RandomGeneratorUtil::nextItem(allowedTargets, *server->getRNG());
 
-
 		auto attackInfo = ca.attackedParts.begin();
 		for ( ; attackInfo != ca.attackedParts.end(); ++attackInfo)
 			if ( attackInfo->attackedPart == target )
 				break;
 
 		if (attackInfo == ca.attackedParts.end()) // new part
 		{
-			CatapultAttack::AttackInfo newInfo{};
-			newInfo.damageDealt = 1;
+			CatapultAttack::AttackInfo newInfo;
+			newInfo.damageDealt = getRandomDamage(server);
 			newInfo.attackedPart = target;
 			newInfo.destinationTile = m->battle()->wallPartToBattleHex(target);
 			ca.attackedParts.push_back(newInfo);
 			attackInfo = ca.attackedParts.end() - 1;
 		}
 		else // already damaged before, update damage
 		{
-			attackInfo->damageDealt += 1;
+			attackInfo->damageDealt += getRandomDamage(server);
 		}
 	}
-
 	server->apply(&ca);
 
+	removeTowerShooters(server, m);
+}
+
+void Catapult::applySeparate(ServerCallback * server, const Mechanics * m, const EffectTarget & target) const
+{
+	assert(!target.empty());
+	auto destination = target.at(0).hexValue;
+	auto desiredTarget = m->battle()->battleHexToWallPart(destination);
+
+	for(int i = 0; i < targetsToAttack; i++)
+	{
+		auto actualTarget = EWallPart::INVALID;
+
+		if ( isWallPartAttackable(desiredTarget, m) &&
+				server->getRNG()->getInt64Range(0, 99)() < getCatapultHitChance(desiredTarget))
+		{
+			actualTarget = desiredTarget;
+		}
+		else
+		{
+			std::vector<EWallPart> potentialTargets = getPotentialTargets(m, false, false);
+
+			if (potentialTargets.empty())
+				break; // everything is gone, can't attack anymore
+
+			actualTarget = *RandomGeneratorUtil::nextItem(potentialTargets, *server->getRNG());
+		}
+		if(actualTarget == EWallPart::INVALID)
+			return; //Effect is wasted
```

Comment:
Is this from old code?
How is this possible?
If we fail to hit intended target, then code always selects random target from permitted.
And if we run out of permitted targets then this line won't be reached due to `break` above

Reply:
It is from copy-paste from server callback. I can remove it, if this will not happen.

Reply:
Hmm.. I would rather add assert to make sure, in case if I am missing something

Reply:
replaced with assert